File: Frontend/shop/src/api/admins.js
export async function fetchAdmins() {
  const response = await fetch("http://localhost:8000/admins/");
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Ошибка при получении списка администраторов");
  }
  return response.json();
}

export async function searchUsersByEmail(emailQuery) {
  const response = await fetch(`http://localhost:8000/admins/search-users?email_query=${encodeURIComponent(emailQuery)}`);
  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Ошибка при поиске пользователей");
  }
  return response.json();
}

export async function promoteToAdmin(userId) {
  const response = await fetch(`http://localhost:8000/admins/promote/${userId}`, {
    method: "POST",
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Ошибка при назначении администратором");
  }

  return response.json();
}

export async function removeAdminRole(userId) {
  const response = await fetch(`http://localhost:8000/admins/remove/${userId}`, {
    method: "POST",
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || "Ошибка при отзыве прав администратора");
  }

  return response.json();
}

File: backend/app/routes/admins.py
from fastapi import APIRouter, HTTPException
from app.repositories.users_repo import (
    get_all_admins,
    search_users_by_email,
    promote_to_admin,
    remove_admin_role
)

router = APIRouter(prefix="/admins")

@router.get("/")
def list_admins():
    """
    Получает список всех администраторов
    """
    try:
        admins = get_all_admins()
        return admins
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка при получении списка администраторов: {str(e)}")

@router.get("/search-users")
def search_users(email_query: str):
    """
    Ищет пользователей по email
    
    Args:
        email_query: Часть email для поиска
    """
    if not email_query or len(email_query.strip()) < 2:
        raise HTTPException(status_code=400, detail="Минимум 2 символа для поиска")
    
    try:
        users = search_users_by_email(email_query.strip())
        return users
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка при поиске пользователей: {str(e)}")

@router.post("/promote/{user_id}")
def promote_user_to_admin(user_id: int):
    """
    Назначает пользователя администратором
    
    Args:
        user_id: ID пользователя
    """
    try:
        user = promote_to_admin(user_id)
        if not user:
            raise HTTPException(status_code=404, detail="Пользователь не найден")
        return user
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка при назначении администратором: {str(e)}")

@router.post("/remove/{user_id}")
def remove_admin(user_id: int):
    """
    Лишает пользователя прав администратора
    
    Args:
        user_id: ID пользователя
    """
    try:
        user = remove_admin_role(user_id)
        if not user:
            raise HTTPException(status_code=404, detail="Администратор не найден или уже не является администратором")
        return {"message": "Права администратора успешно отозваны", "user": user}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Ошибка при отзыве прав администратора: {str(e)}")
